### PR TITLE
Optimize file_tree ext_pillar and update file.managed to allow for binary contents

### LINKED ISSUE #9569: How can a binary file, such as a license key, be distributed via Pillar?
There doesn't seem to be a way to provision binary data into pillar values in order to feed to file.contents_pillar.

### PR DESCRIPTION
**Pull request by terminalmage**
The contents/contents_pillar/contents_grains options assume that the contents
are string data, this pull request allows binary contents to be deployed.
Additionally, some changes have been made to file_tree to make reading in each
file more efficient. Finally, a usage example has been added to the FAQ.

Resolves #9569.

### CODE REVIEW

**Comment by cachedout** [comment]
@terminalmage Could you please rebase and fix some lint? Thanks.

**Comment by terminalmage** [comment]
@cachedout Still a lot of test failures, are these familiar? They don't seem related, but I'm looking deeper into it now.

File: salt/modules/file.py [hunk 1]
```diff
@@ -1656,21 +1656,28 @@ def replace(path,
     '''
     .. versionadded:: 0.17.0
 
-    Replace occurrences of a pattern in a file
+    Replace occurrences of a pattern in a file. If ``show_changes`` is
+    ``True``, then a diff of what changed will be returned, otherwise a
+    ``True`` will be returnd when changes are made are made, and ``False`` when
```
Comment (iggy) [inline salt/modules/file.py]: `are made are made`

Comment (terminalmage) [inline salt/modules/file.py]: Thanks!

**Comment by cachedout** [comment]
@terminalmage That was my bad. A rebase wouldn't have fixed this. Apologies.

**Comment by cachedout** [comment]
@terminalmage A little lint to be fixed up here, please.

**Pull request merged**